In [1]:
import pandas as pd
import numpy as np

In [4]:
pd.set_option('display.max_columns', None)

In [5]:
df = pd.read_csv('../data/LA/Crime_Data_from_2010_to_Present.csv',
                 header=0)
print('df.shape:', df.shape)
df.head()

df.shape: (1906621, 26)


DR Number Date Reported Date Occurred  Time Occurred  Area ID Area Name  \
0  171305008    01/18/2017    01/18/2017            130       13    Newton   
1  171305151    01/20/2017    01/20/2017           1300       13    Newton   
2  171305393    01/24/2017    01/23/2017           1900       13    Newton   
3  171305398    01/24/2017    01/24/2017            800       13    Newton   
4  171306365    02/08/2017    02/07/2017           2000       13    Newton   

   Reporting District  Crime Code Crime Code Description MO Codes  Victim Age  \
0                1309         510       VEHICLE - STOLEN      NaN        16.0   
1                1341         510       VEHICLE - STOLEN      NaN        16.0   
2                1345         510       VEHICLE - STOLEN      NaN        16.0   
3                1353         510       VEHICLE - STOLEN      NaN        16.0   
4                1333         510       VEHICLE - STOLEN      NaN        16.0   

  Victim Sex Victim Descent  Premise Code Premise Description  \
0        NaN            NaN         108.0         PARKING LOT   
1        NaN            NaN         101.0              STREET   
2        NaN            NaN         101.0              STREET   
3        NaN            NaN         101.0              STREET   
4        NaN            NaN         101.0              STREET   

   Weapon Used Code Weapon Description Status Code Status Description  \
0               NaN                NaN          IC        Invest Cont   
1               NaN                NaN          IC        Invest Cont   
2               NaN                NaN          IC        Invest Cont   
3               NaN                NaN          IC        Invest Cont   
4               NaN                NaN          IC        Invest Cont   

   Crime Code 1  Crime Code 2  Crime Code 3  Crime Code 4  \
0         510.0           NaN           NaN           NaN   
1         510.0           NaN           NaN           NaN   
2         510.0           NaN           NaN           NaN   
3         510.0           NaN           NaN           NaN   
4         510.0           NaN           NaN           NaN   

                                   Address                     Cross Street  \
0   700 S  SANTA FE                     AV                              NaN   
1                                 BROADWAY  37TH                         ST   
2  3500    HOOPER                       AV                              NaN   
3                                   VERNON                         MCKINLEY   
4   800 E  32ND                         ST                              NaN   

              Location   
0    (34.0344, -118.23)  
1  (34.0158, -118.2759)  
2   (34.012, -118.2521)  
3  (34.0039, -118.2609)  
4  (34.0163, -118.2624)

### Cleanup data

In [14]:
df.isnull().sum()

DR Number                       0
Date Reported                   0
Date Occurred                   0
Time Occurred                   0
Area ID                         0
Area Name                       0
Reporting District              0
Crime Code                      0
Crime Code Description        361
MO Codes                   205102
Victim Age                 150186
Victim Sex                 175671
Victim Descent             175712
Premise Code                   85
Premise Description          3229
Weapon Used Code          1270353
Weapon Description        1270354
Status Code                     2
Status Description              0
Crime Code 1                    8
Crime Code 2              1783379
Crime Code 3              1903637
Crime Code 4              1906529
Address                         0
Cross Street              1588402
Location                        9
dtype: int64

### There are only 361 rows (out of almost 2 million), with values in 'Crime Code Description' that are null. Remove those rows

In [17]:
df = df.dropna(axis=0, how='any', subset=['Crime Code Description'])
print('df.shape:', df.shape)
df.head()

df.shape: (1906260, 26)


DR Number Date Reported Date Occurred  Time Occurred  Area ID Area Name  \
0  171305008    01/18/2017    01/18/2017            130       13    Newton   
1  171305151    01/20/2017    01/20/2017           1300       13    Newton   
2  171305393    01/24/2017    01/23/2017           1900       13    Newton   
3  171305398    01/24/2017    01/24/2017            800       13    Newton   
4  171306365    02/08/2017    02/07/2017           2000       13    Newton   

   Reporting District  Crime Code Crime Code Description MO Codes  Victim Age  \
0                1309         510       VEHICLE - STOLEN      NaN        16.0   
1                1341         510       VEHICLE - STOLEN      NaN        16.0   
2                1345         510       VEHICLE - STOLEN      NaN        16.0   
3                1353         510       VEHICLE - STOLEN      NaN        16.0   
4                1333         510       VEHICLE - STOLEN      NaN        16.0   

  Victim Sex Victim Descent  Premise Code Premise Description  \
0        NaN            NaN         108.0         PARKING LOT   
1        NaN            NaN         101.0              STREET   
2        NaN            NaN         101.0              STREET   
3        NaN            NaN         101.0              STREET   
4        NaN            NaN         101.0              STREET   

   Weapon Used Code Weapon Description Status Code Status Description  \
0               NaN                NaN          IC        Invest Cont   
1               NaN                NaN          IC        Invest Cont   
2               NaN                NaN          IC        Invest Cont   
3               NaN                NaN          IC        Invest Cont   
4               NaN                NaN          IC        Invest Cont   

   Crime Code 1  Crime Code 2  Crime Code 3  Crime Code 4  \
0         510.0           NaN           NaN           NaN   
1         510.0           NaN           NaN           NaN   
2         510.0           NaN           NaN           NaN   
3         510.0           NaN           NaN           NaN   
4         510.0           NaN           NaN           NaN   

                                   Address                     Cross Street  \
0   700 S  SANTA FE                     AV                              NaN   
1                                 BROADWAY  37TH                         ST   
2  3500    HOOPER                       AV                              NaN   
3                                   VERNON                         MCKINLEY   
4   800 E  32ND                         ST                              NaN   

              Location   
0    (34.0344, -118.23)  
1  (34.0158, -118.2759)  
2   (34.012, -118.2521)  
3  (34.0039, -118.2609)  
4  (34.0163, -118.2624)

In [23]:
df['Crime Code Description'].unique()

array(['VEHICLE - STOLEN', 'PICKPOCKET',
       'THEFT PLAIN - PETTY ($950 & UNDER)', 'THEFT OF IDENTITY',
       'INDECENT EXPOSURE',
       'THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)', 'BURGLARY',
       'VANDALISM - FELONY ($400 & OVER, ALL CHURCH VANDALISMS) 0114',
       'ARSON', 'TRESPASSING',
       'THEFT FROM MOTOR VEHICLE - GRAND ($400 AND OVER)', 'PEEPING TOM',
       'BIKE - STOLEN', 'VIOLATION OF COURT ORDER', 'CONTEMPT OF COURT',
       'BOMB SCARE', 'OTHER MISCELLANEOUS CRIME', 'BUNCO, PETTY THEFT',
       'BURGLARY FROM VEHICLE', 'BATTERY - SIMPLE ASSAULT',
       'VANDALISM - MISDEAMEANOR ($399 OR UNDER)', 'STALKING',
       'DISTURBING THE PEACE', 'LETTERS, LEWD',
       'VIOLATION OF TEMPORARY RESTRAINING ORDER', 'THEFT, PERSON',
       'RAPE, ATTEMPTED', 'BRANDISH WEAPON', 'BURGLARY, ATTEMPTED',
       'ILLEGAL DUMPING', 'CRIMINAL THREATS - NO WEAPON DISPLAYED',
       'DOCUMENT FORGERY / STOLEN FELONY',
       'ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT',

In [18]:
df['Crime Code Description'].nunique()

149

In [57]:
crime_description = pd.Series(df['Crime Code Description'].unique())
crime_description[crime_description.str.contains('KID')].values

array(['KIDNAPPING', 'KIDNAPPING - GRAND ATTEMPT'], dtype=object)

In [89]:
SEVERITY_HIGH = 100
SEVERITY_MEDIUM = 50
SEVERITY_LOW  = 20

def str_contains_any_substr(string, substrings):
    for subs in substrings:
        if subs in string:
            return True
    return False

havent_handled = []
def crime_severity(x):
    """
    Converts the primary type and description into a crime severity value.
    This value is based on how much a particular crime would affect a user
    that is either:
    1. Renting short-term and wants to roam around the city
    2. Renting long-term and wants to find a good neighborhood to roam in
    """
    description = x
    if ('VEH' in description):
        if 'THEFT FROM' in description or \
            'BURGLARY FROM' in description:
            return SEVERITY_LOW
        elif 'STOLEN' in description:
            return SEVERITY_MEDIUM
        elif 'THROWING' in description or \
             'SHOTS FIRED' in description:
            return SEVERITY_HIGH
              
    if 'THEFT' in description:
        if 'PETTY' in description or \
            'DEFRAUD' in description:
            return SEVERITY_LOW
        elif 'GRAND' in description:
            return SEVERITY_MEDIUM
        elif 'PERSON' in description:
            return SEVERITY_HIGH
        else:
            return SEVERITY_LOW
    
    if str_contains_any_substr(description,
                              ['BLOCKING', 'BIKE',
                               'UNAUTHORIZED COMPUTER ACCESS',
                               'BRIBERY',
                               'TRESPASSING', 
                               'OTHER MISCELLANEOUS CRIME',
                               'FALSE IMPRISONMENT',
                               'ORAL COPULATION',
                               'TILL TAP - PETTY ($950 & UNDER)',
                               'TILL TAP - ATTEMPT', 
                               'SHOPLIFTING - ATTEMPT',
                               'DOCUMENT WORTHLESS ($200.01 & OVER)',
                               'DOCUMENT WORTHLESS ($200 & UNDER)',
                               'CONTRIBUTING', 
                               'BIGAMY',
                              ]):
        return SEVERITY_LOW
    
    if str_contains_any_substr(description, 
                               ['PICKPOCKET', 'CREDIT CARDS',
                                'DRUNK ROLL', 'FORGERY',
                                'TELEPHONE', 'BURGLARY', 'ROBBERY',
                                'PURSE SNATCHING', 'BUNCO',
                                'INDECENT EXPOSURE', 
                                'CONTEMPT OF COURT', 
                                'COUNTERFEIT', 
                                'FAILURE TO YIELD',
                                'FAILURE TO DISPERSE',
                                'DRIVING WITHOUT OWNER CONSENT (DWOC)',
                                'CRUELTY TO ANIMALS',
                                'DISRUPT SCHOOL',                                
                               ]):
        return SEVERITY_MEDIUM
    
    if 'VANDALISM - FELONY' in description:
        return SEVERITY_MEDIUM
    elif 'VANDALISM - MISDEAMEANOR' in description:
        return SEVERITY_LOW
    
    if 'CHILD' in description:
        if 'ANNOYING' in description:
            return SEVERITY_LOW
        elif 'SIMPLE' in description:
            return SEVERITY_MEDIUM
        else:
            return SEVERITY_HIGH
        
    if 'SEX' in description:
        if 'INC MUTUAL CONSENT' in description:
            return SEVERITY_LOW
        else:
            return SEVERITY_HIGH
    
    if 'WEAPON' in description:
        if 'NO WEAPON DISPLAYED' in description:
            return SEVERITY_LOW
        else:
            return SEVERITY_MEDIUM
    
    if str_contains_any_substr(description,
                               ['DISCHARGE FIREARMS', 'SHOTS FIRED',
                                'FIREARMS RESTRAINING ORDER']):
        return SEVERITY_HIGH
    elif 'REPLICA FIREARMS' in description:
        return SEVERITY_MEDIUM
    
    if str_contains_any_substr(description,
                                ['RAPE', 'CHILD PORNOGRAPHY',
                                 'BATTERY', 'HUMAN TRAFFICKING',
                                 'KIDNAPPING', 'ASSAULT',
                                 'CRM AGNST CHLD',
                                 'ARSON', 
                                 'PEEPING TOM',
                                 'PROWLER', 
                                 'PIMPING',
                                 'VIOLATION OF COURT ORDER', 
                                 'BOMB SCARE',
                                 'STALKING',
                                 'DISTURBING THE PEACE',
                                 'LETTERS, LEWD', 
                                 'VIOLATION OF TEMPORARY RESTRAINING ORDER',
                                 'ILLEGAL DUMPING', 
                                 'RESISTING ARREST',
                                 'VIOLATION OF RESTRAINING ORDER', 
                                 'FALSE POLICE REPORT',
                                 'CRIMINAL HOMICIDE', 
                                 'EXTORTION', 
                                 'PANDERING',
                                 'THREATENING PHONE CALLS/LETTERS', 
                                 'LEWD CONDUCT', 
                                 'RECKLESS DRIVING', 
                                 'ABORTION/ILLEGAL',
                                 'BOAT - STOLEN', 
                                 'INCITING A RIOT', 
                                 'LYNCHING', 
                                 'LYNCHING - ATTEMPTED', 
                                 'CONSPIRACY',
                                 'DRUGS, TO A MINOR', 
                                 'MANSLAUGHTER, NEGLIGENT',
                                 'TRAIN WRECKING',
                                ]):
        return SEVERITY_HIGH
 
    havent_handled.append(description)
    return SEVERITY_LOW

In [90]:
df['Severity'] = df['Crime Code Description'].apply(crime_severity)

In [91]:
hh = pd.Series(pd.Series(havent_handled).unique())

In [92]:
hh.unique()

array([], dtype=float64)

In [94]:
# hh[hh.str.contains('TRESPASS')].values